### DISCUSSION TITLE
Programmatically updating handles doesn't update edges

### DISCUSSION THREAD

**Discussion by gilneto8** [xyflow/xyflow#1151] [Q&A]
Hey guys,

I have an issue where I have a custom node which I can configure with any number of handles. This works perfectly and they are all recognized on the flow after discovering the `useUpdateNodeInternals` hook.
Now I'm facing a bug that I don't actually understand the origin of. As I said, the node is updated successfully and all the handles are recognized, but if that node had edges connected to it, they don't get updated automatically.

As an example, imagine I have this:
![image](https://user-images.githubusercontent.com/15315355/116703516-4ae1e580-a9c2-11eb-92e4-d3a67e8e6582.png)
The two Debug nodes are connected to 2 of the 3 handles of the Join node. I won't need the 3rd handle from Join, so I'll update 
the Join node to have only two handles and, thus, resizing and repositioning it.

![image](https://user-images.githubusercontent.com/15315355/116703675-74027600-a9c2-11eb-91ab-8e522a625644.png)
As you can see, the edges don't update correctly. They are connected internally, but visually they don't appear so. If I try to reconnect one of them again (even if to the same handle), they'll all update and show correctly.

![image](https://user-images.githubusercontent.com/15315355/116703855-a613d800-a9c2-11eb-8706-2bae346dca36.png)

So my question is: am I missing something? Should I update something more when I use the `useUpdateNodeInternals` hook? 

Thanks in advance.

**Comment by moklick** [comment]
This should work when you call `useUpdateNodeInternals` after you changed the positions of the handles.

**Reply by gilneto8** [reply]
I'm using a form to update the number of handles, and I save this number inside the `data` property of the node, so that whenever I'm rendering it I can access it and create the handles dynamically. 

It seems like the edges are only updated when I explicitly call `updateEdge` on **any** edge.

I'll try to create a side project to understand if this is a problem on my project.

**Reply by moklick** [reply]
When are you calling `useUpdateNodeInternals` ?

**Reply by gilneto8** [reply]
The sequence of actions is as follows:

1. The form is updated and the saving action is called;
2. On save, the node data property is updated with the new values (in this case, #handles);
3. On my custom node component, I have a `useEffect` hook that's fired whenever a new set of handles is saved. Here, the handles are created and saved as the 'current configuration' to be rendered;
4. Afterwards, I call the `updateNodeInternals` hook.

The code for these last 2 steps is as follows:

```javascript
  const [inputs, setInputs] = useState<Array<() => JSX.Element>>(renderHandles());
  useEffect(() => {
    setInputs(renderHandles());
    if (selected) updateNodeInternals(selected.id);
  }, [data]);
  ```

And the code to create the handles:

```javascript
const renderHandles = (): Array<() => JSX.Element> => {
    let handles: Array<() => JSX.Element> = [];
    const difference = 100 / (Number(data.numberOfInputs) + 1);
    let leftOffset: number = difference;
    for (let i = 0; i < data.numberOfInputs; i++) {
      const currentOffset = leftOffset;
      handles.push(() => (
        <Handle
          id={`join-in-${i + 1}`}
          type="target"
          className={'handle'}
          position={Position.Top}
          style={{ left: `${currentOffset.toFixed()}%` }}
          isValidConnection={methods.isValidConnection}
        />
      ));
      leftOffset += difference;
    }
    return handles;
  };
```

Unless I must account for some asynchronicity, I'm not sure where this can fail.

**Reply by gilneto8** [reply]
(Bear in mind that the `selected` variable is not undefined at the time of the hook calling, but I have to confirm that previously because I'm using Typescript and the compiler doesn't like that)

**Reply by moklick** [reply]
why are you putting `renderHandles` inside the local state and not just returning it in the render function? With this setup `updatNodeInternals` gets called before the handles are rendered.

**Reply by gilneto8** [reply]
You're right, I've added too much unnecessary complexity. I've removed the local state variables and added the `renderHandles` method directly to the `render` function. Unfortunately, the result is the same.

But i think this might have to do with asynchronicity, now that we're looking at the right parts. I've noticed that after I force the `updateNodeInternals` to be called after the `renderHandles` call, any action I do afterwards will reset the edges to the correct state. It's manual, but it is happening, which is a good thing. And it also indicates that I'm doing something wrong, so I guess I'm in the right direction.

This is the `render` function right now, that's almost working:

```javascript
return (
    <div
      className={`rendered-node ${selected?.id === id ? 'node-selected' : hovering ? 'node-hovering' : ''}`}
      style={{ width: `${25 * data.numberOfInputs}px` }}
      onMouseEnter={() => setHovering(true)}
      onMouseLeave={() => setHovering(false)}
    >
      <NodeHeader nodeType={props.type} color={NodeCategories.Misc.color} />
      <div id={'content'}>
        {interactive && selected?.id === id && <Cross badge pos={{ x: 7, y: 7 }} onClick={() => methods.removeNode(id)} />}
        {renderHandles()}
        {selected && updateNodeInternals(selected.id)}
        <div className={'label'}>{data.label}</div>
        <Handle
          id={'join-out'}
          className={'handle'}
          type="source"
          position={Position.Bottom}
          isValidConnection={methods.isValidConnection}
        />
      </div>
    </div>
  );
  ```
  
  I'll have to keep investigating, but at least you pointed me in the right direction. Thank you very much.

**Reply by moklick** [reply]
This looks better :) Does it work when you put this in your custom node?

```js 
useEffect(() => updateNodeInternals(id), [data.numberOfInputs]);
```

**Reply by gilneto8** [reply]
That doesn't change a thing, unfortunately. It's still as on the previous scenario, which is not the worst case, but it's not automatic either. I'll have to investigate how the calls are stacked and when I should actually call the hook.
